sky/provision/gcp/config.py

```diff
@@ -0,0 +1,720 @@
+
+    for _ in range(MAX_POLLS):
+        result = crm.operations().get(name=operation['name']).execute()
+        if 'error' in result:
+            raise Exception(result['error'])
+
+        if 'done' in result and result['done']:
+            logger.info('wait_for_crm_operation: Operation done.')
+            break
+
+        time.sleep(POLL_INTERVAL)
+
+    return result
+
+
+def wait_for_compute_global_operation(project_name, operation, compute):
+    """Poll for global compute operation until finished."""
+    logger.info('wait_for_compute_global_operation: '
+                'Waiting for operation {} to finish...'.format(
+                    operation['name']))
+
+    for _ in range(MAX_POLLS):
+        result = (compute.globalOperations().get(
+            project=project_name,
+            operation=operation['name'],
+        ).execute())
+        if 'error' in result:
+            raise Exception(result['error'])
+
+        if result['status'] == 'DONE':
+            logger.info('wait_for_compute_global_operation: Operation done.')
+            break
+
+        time.sleep(POLL_INTERVAL)
+
+    return result
+
+
+def _create_crm(gcp_credentials=None):
+    return gcp.build('cloudresourcemanager',
+                     'v1',
+                     credentials=gcp_credentials,
+                     cache_discovery=False)
+
+
+def _create_iam(gcp_credentials=None):
+    return gcp.build('iam',
+                     'v1',
+                     credentials=gcp_credentials,
+                     cache_discovery=False)
+
+
+def _create_compute(gcp_credentials=None):
+    return gcp.build('compute',
+                     'v1',
+                     credentials=gcp_credentials,
+                     cache_discovery=False)
+
+
+def _create_tpu(gcp_credentials=None):
+    return gcp.build(
+        'tpu',
+        TPU_VERSION,
+        credentials=gcp_credentials,
+        cache_discovery=False,
+        discoveryServiceUrl='https://tpu.googleapis.com/$discovery/rest',
+    )
+
+
+def construct_clients_from_provider_config(provider_config):
+    """
+    Attempt to fetch and parse the JSON GCP credentials from the provider
+    config yaml file.
+
+    tpu resource (the last element of the tuple) will be None if
+    `_has_tpus` in provider config is not set or False.
+    """
+    gcp_credentials = provider_config.get('gcp_credentials')
+    if gcp_credentials is None:
+        logger.debug('gcp_credentials not found in cluster yaml file. '
+                     'Falling back to GOOGLE_APPLICATION_CREDENTIALS '
+                     'environment variable.')
+        tpu_resource = (_create_tpu() if provider_config.get(
+            HAS_TPU_PROVIDER_FIELD, False) else None)
+        # If gcp_credentials is None, then discovery.build will search for
+        # credentials in the local environment.
+        return _create_crm(), _create_iam(), _create_compute(), tpu_resource
+
```

Comment:
```suggestion
# Note: The following code has not been used yet, as we will never set `gcp_credentials`
# in provider_config. It will only be used when we allow users to specify
# their own credeitals.
```